src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion_controlnet.py

```diff
@@ -0,0 +1,375 @@
+# limitations under the License.
+from dataclasses import dataclass
+from typing import List, Optional, Union
+
+import numpy as np
+import PIL
+from ...models import AutoencoderKL, UNet2DConditionModel, ControlNetModel
+from .modeling_ctx_clip import ContextCLIPTextModel
+from transformers import CLIPTokenizer
+from .. import DiffusionPipeline
+import torch
+from ...schedulers import PNDMScheduler
+from ...utils import (
+    BaseOutput,
+    is_accelerate_available,
+    is_accelerate_version,
+    logging,
+    randn_tensor,
+    replace_example_docstring,
+)
+from ...utils.pil_utils import PIL_INTERPOLATION
+from torch import nn
+from transformers.activations import QuickGELUActivation as QuickGELU
+from .modeling_blip2 import Blip2QFormerModel
+import tqdm
+from torchvision import transforms
+from torchvision.transforms.functional import InterpolationMode
+from .blip_image_processing import BlipImageProcessor
+from PIL import Image
+from ..pipeline_utils import ImagePipelineOutput
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+import re
+
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```py
+        >>> from diffusers import BlipDiffusionControlNetPipeline
+        >>> from PIL import Image
+        >>> from diffusers.utils import load_image
+
+        >>> blip_diffusion_pipe = BlipDiffusionPipeline.from_pretrained('ayushtues/blipdiffusion-controlnet')
+        >>> blip_diffusion_pipe.to('cuda')
+
+        >>> style_subject = ["flower"]
+        >>> tgt_subject = ["teapot"]
+        >>> text_prompt = ["on a marble table"]
+        >>> cldm_cond_image = load_image(
+        ...     "https://huggingface.co/datasets/ayushtues/blipdiffusion_images/resolve/main/kettle.jpg"
+        ... )
+        >>> style_image = load_image(
+        ...     "https://huggingface.co/datasets/ayushtues/blipdiffusion_images/resolve/main/flower.jpg"
+        ... )
+        >>> iter_seed = 88888
+        >>> guidance_scale = 7.5
+        >>> num_inference_steps = 50
+        >>> negative_prompt = "over-exposure, under-exposure, saturated, duplicate, out of frame, lowres, cropped, worst quality, low quality, jpeg artifacts, morbid, mutilated, out of frame, ugly, bad anatomy, bad proportions, deformed, blurry, duplicate"
+
+        
+        >>> output = blip_diffusion_pipe(
+        ...     text_prompt,
+        ...     style_image,
+        ...     style_subject,
+        ...     tgt_subject,
+        ...     cldm_cond_image,
+        ...     seed=iter_seed + i,
+        ...     guidance_scale=guidance_scale,
+        ...     num_inference_steps=num_inference_steps,
+        ...     neg_prompt=negative_prompt,
+        ...     height=512,
+        ...     width=512,
+        ...     )
+        >>> output[0][0].save("kettle_new.png")
+        ```
+"""
+
+
+class BlipDiffusionControlNetPipeline(DiffusionPipeline):
+    """
+    Pipeline for Canny Edge based Controlled subject-driven generation using Blip Diffusion.
```

Comment:
If a user loads a ControlNet trained on say, depth conditions, I think we should be able to handle that here too, no? 

I think for that to work successfully, we need to delegate the preprocessing steps (for the conditioning image as needed by a ControlNet) to the user and then it should work right off the bat. This delegation is absolutely fine because we already do it for our existing ControlNets. 

What are your thoughts? 

This way, I think we open various doors of very interesting applications.

Reply:
Yeah makes sense, let me refactor and try to add 2 different controlnet examples

Reply:
Done